### PR TITLE
Fix: support bidirectional sliding window attention in torch native backend

### PR DESCRIPTION
**Pull request by approximated-intelligence**
The `TorchNativeAttnBackend` previously hardcoded sliding window masks  to be causal (`k_pos <= q_pos`) and gated the feature behind `if causal`.  This caused encoder models (like ModernBERT) to ignore  sliding window attention, falling back to full global attention and  breaking long-context parity with HuggingFace.

This introduces a `causal` flag to `_make_sliding_window_mask`  (defaulting to `True` for backward compatibility). When `False`, it  generates a correct bidirectional window mask 
(`torch.abs(q_pos - k_pos) <= sliding_window_size`). It also removes  the `causal` requirement in `forward_extend` so encoders can properly  utilize local attention.

## Motivation

The `TorchNativeAttnBackend` previously hardcoded sliding window masks to be causal (`k_pos <= q_pos`) and gated the sliding window feature behind `if causal`. 

This caused encoder models (which use `causal=False`) to completely ignore sliding window attention. The backend silently fell back to full global attention, severely corrupting long-context embeddings and breaking parity with HuggingFace Transformers.

This fix is required to enable the upcoming ModernBERT model architecture (development branch: [approximated-intelligence/sglang:modernbert](https://github.com/approximated-intelligence/sglang/blob/modernbert/python/sglang/srt/models/modernbert.py)) to function correctly.

## Modifications

**`python/sglang/srt/layers/attention/torch_native_backend.py`**
- Introduced a `causal: bool = True` parameter to `_make_sliding_window_mask` (defaulting to `True` to preserve backward compatibility for causal LLMs).
- When `causal=False`, the mask generator now creates a correct bidirectional window mask: `torch.abs(q_pos - k_pos) <= sliding_window_size`.
- Passed the `causal` flag down from `_run_sdpa_forward_extend` and `_run_sdpa_forward_decode`.
- Removed the `causal and ...` requirement in `forward_extend` so that encoder models can properly utilize local attention.

## Accuracy Tests

The following test script was used to compare SGLang HTTP outputs against HuggingFace `AutoModel` outputs on CPU (`--device cpu --dtype float32`), testing an encoder model that utilizes bidirectional sliding window attention. 

<details>
<summary>test_embeddings.py</summary>

```python
#!/usr/bin/env python3
"""
Compare ModernBERT embeddings: Transformers vs SGLang HTTP endpoint.
Tests short (single/batch) and long context (sliding window).
Assumes server is already running at http://localhost:30000 (--is-embedding).
"""

import random
import torch
import requests
from transformers import AutoTokenizer, AutoModel

import logging
# Suppress HuggingFace load reports and warnings
logging.getLogger("transformers").setLevel(logging.ERROR) 

MODEL_NAME = "jhu-clsp/ettin-encoder-17m"
SERVER_URL = "http://localhost:30000/v1/embeddings"

SINGLE = "Hello world"
BATCH = [
    "The quick brown fox jumps over the lazy dog",
    "SGLang is a fast serving framework",
    "ModernBERT uses RoPE and GeGLU",
]

# High-perplexity, highly diverse text to prevent tokenizer compression
# and stress the attention mechanism with varied distributions.
DIVERSE_SENTENCES = [
    "The mitochondrion is the powerhouse of the cell, generating ATP through oxidative phosphorylation.",
    "Beneath the varnished surface of the mahogany table, an intricate network of termites had compromised the structural integrity.",
    "SGLang leverages a radix tree for managing the KV cache, ensuring efficient memory reuse across shared prefixes.",
    "Quasars emit electromagnetic radiation across the spectrum, from radio waves to gamma rays.",
    "The quick brown fox jumps over the lazy dog, but the dog was actually a shapeshifting alien.",
    "ModernBERT utilizes Rotary Position Embeddings and a sliding window attention mechanism to handle long sequences efficiently.",
    "In the depths of the Mariana Trench, bioluminescent organisms thrive under crushing hydrostatic pressure.",
    "Algorithmic trading relies on low-latency execution to capitalize on microsecond arbitrage opportunities.",
    "The enigmatic manuscript, written in an unknown script, baffled cryptographers for centuries.",
    "She whispered a forgotten incantation, and the obsidian obelisk began to levitate.",
    "Neural networks with GeGLU activations often outperform standard ReLU-based counterparts in language modeling.",
    "The migration of monarch butterflies spans thousands of miles, guided by an innate magnetoreception.",
    "Atmospheric carbon dioxide levels continue to rise, accelerating the greenhouse effect.",
    "He navigated the bustling bazaar, haggling fiercely over the price of saffron and cardamom.",
    "Quantum entanglement suggests that particles can be inextricably linked regardless of distance.",
    "The Byzantine general's problem illustrates the difficulty of achieving consensus in distributed systems.",
    "A sudden squall swept across the loch, churning the placid water into a frothy tempest.",
    "CRISPR-Cas9 technology allows for precise genomic editing, revolutionizing personalized medicine.",
    "The ancient Sumerians invented cuneiform script to record administrative transactions on clay tablets.",
    "In machine learning, gradient descent optimization can sometimes get trapped in local minima.",
    "The majestic sequoia trees of California can live for thousands of years.",
    "A Gordian knot of bureaucratic red tape delayed the infrastructure project indefinitely.",
    "Photosynthesis converts solar energy into chemical energy, sustaining the foundation of most ecosystems.",
    "The harpsichordist played a delicate fugue by Johann Sebastian Bach.",
    "Vector databases enable efficient similarity search for high-dimensional embedding spaces.",
    "The precipitation in the Amazon rainforest is heavily influenced by transpiration from the dense canopy.",
    "A rogue planet, ejected from its solar system, wandered through the interstellar void.",
    "The alchemist sought the philosopher's stone, hoping to transmute lead into gold.",
    "FlashAttention optimizes the memory hierarchy by minimizing high-bandwidth memory reads and writes.",
    "The erratic flight pattern of the bat was captured by the high-speed camera.",
    "Symbiotic relationships, such as that between clownfish and sea anemones, are vital for ecosystem stability.",
    "Despite the freezing temperatures, the expedition continued their ascent towards the Himalayan summit.",
    "The cryptography algorithm utilized a 4096-bit key to ensure absolute data security.",
    "Economic indicators showed a surprising rebound in the manufacturing sector during the third quarter.",
    "The orchestral conductor dramatically slowed the tempo for the final movement.",
]

def generate_long_context(tokenizer, target_tokens=1800):
    """Generates a high-perplexity long context by shuffling diverse sentences."""
    sentences = DIVERSE_SENTENCES.copy()
    random.seed(42) # Deterministic but shuffled
    long_text = ""
    
    while True:
        random.shuffle(sentences)
        long_text += " ".join(sentences) + " "
        token_count = len(tokenizer.encode(long_text, add_special_tokens=True))
        if token_count >= target_tokens:
            return long_text, token_count


def embed_transformers(texts):
    tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
    model = AutoModel.from_pretrained(MODEL_NAME).eval()
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model.to(device)

    # Allow up to model's max position embeddings (7999 in config)
    inputs = tokenizer(texts, padding=True, truncation=True, max_length=8192, return_tensors="pt").to(device)
    with torch.no_grad():
        hidden = model(**inputs).last_hidden_state
        mask = inputs["attention_mask"].unsqueeze(-1).float()
        pooled = (hidden * mask).sum(dim=1) / mask.sum(dim=1).clamp(min=1e-9)
        pooled = torch.nn.functional.normalize(pooled, p=2, dim=1)
    return pooled.cpu()


def embed_server(texts):
    payload = {"input": texts, "model": MODEL_NAME}
    resp = requests.post(SERVER_URL, json=payload)
    resp.raise_for_status()
    data = resp.json()["data"]
    indexed = {item["index"]: torch.tensor(item["embedding"]) for item in data}
    return torch.stack([indexed[i] for i in range(len(texts))])


def compare(tf_emb, srv_emb, label):
    cos = torch.nn.functional.cosine_similarity(tf_emb, srv_emb, dim=0).item()
    mse = torch.nn.functional.mse_loss(tf_emb, srv_emb).item()
    status = "✅ PASS" if cos > 0.995 else "❌ FAIL"
    print(f"  {status} | CosSim={cos:.6f} | MSE={mse:.8f} | {label}")


def run_test(texts, label):
    print(f"\n--- {label} ---")
    try:
        tf_emb = embed_transformers(texts)
        srv_emb = embed_server(texts)
    except Exception as e:
        print(f"  Error: {e}")
        return

    if len(texts) == 1:
        compare(tf_emb[0], srv_emb[0], label)
    else:
        for i in range(len(texts)):
            compare(tf_emb[i], srv_emb[i], f"{label} [{i}]")


if __name__ == "__main__":
    print("Computing embeddings (Transformers vs SGLang)...")
    
    # 1. Short context
    run_test([SINGLE], "Single")
    run_test(BATCH, "Batch")
    
    # 2. Long context (tests sliding window with high perplexity)
    print("\nGenerating high-perplexity long context...")
    tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
    long_text, token_count = generate_long_context(tokenizer, target_tokens=1750)
    print(f"Long context length: {token_count} tokens")
    run_test([long_text], "Long Context (Sliding Window)")
    longer_text, token_count = generate_long_context(tokenizer, target_tokens=3000)
    print(f"Long context length: {token_count} tokens")
    run_test([longer_text], "Longer Context (Sliding Window)")

    print("\nAll comparisons done.")
```
</details>

**Server Launch Command:**
```bash
python -m sglang.launch_server --device cpu --model-path jhu-clsp/ettin-encoder-17m --is-embedding --dtype float32 --port 30000 --disable-radix-cache --chunked-prefill-size -1
```

**Results:**
```text
--- Single ---
  ✅ PASS | CosSim=1.000000 | MSE=0.00000000 | Single
 
--- Batch ---
  ✅ PASS | CosSim=1.000000 | MSE=0.00000000 | Batch [0]
  ✅ PASS | CosSim=1.000000 | MSE=0.00000000 | Batch [1]
  ✅ PASS | CosSim=1.000000 | MSE=0.00000000 | Batch [2]
 
--- Long Context (Sliding Window) ---
  ✅ PASS | CosSim=1.000000 | MSE=0.00000000 | Long Context (1938 tokens)

--- Longer Context (Sliding Window) ---
  ✅ PASS | CosSim=1.000000 | MSE=0.00000000 | Longer Context (3228 tokens)
```

## Speed Tests and Profiling

N/A. This PR fixes a numerical accuracy bug in the CPU attention backend. No regressions are expected for existing causal models, as the mask generation changes default to the old causal behavior.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #31430724127](https://github.com/sgl-project/sglang/actions/runs/31430724127)
Latest PR Test (Extra): :x: [Run #31430724002](https://github.com/sgl-project/sglang/actions/runs/31430724002)

### CODE REVIEW

**Comment by approximated-intelligence** [comment]
some tests are gated, probably this PR needs the label `run-ci`
